wikipedia.go

```diff
@@ -0,0 +1,125 @@
+package meteo
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+)
+
+const (
+	maxRawResults = 10
+)
+
+type wikipediaPlaces struct {
+	Geonames []struct {
+		Lat         float64 `json:"lat"`
+		Lng         float64 `json:"lng"`
+		CountryCode string  `json:"countryCode"`
+		Title       string  `json:"title"`
+	} `json:"geonames"`
+}
+
+type wikipediaOption func(wk *WikipediaClient) error
+
+// WithWikiUserAgent returns a func option for
+// setting up a custom User-Agent used in
+// http requests.
+func WithWikiUserAgent(ua string) wikipediaOption {
+	return func(wk *WikipediaClient) error {
+		wk.UA = ua
+		return nil
+	}
+}
+
+// WithWikiHTTPClient returns a func option for
+// configuring a custom http client.
+func WithWikiHTTPClient(hc *http.Client) wikipediaOption {
+	return func(wk *WikipediaClient) error {
+		wk.HTTPClient = hc
+		return nil
+	}
+}
+
+// WikipediaClient implements Nameresolver interface.
+type WikipediaClient struct {
+	UA         string
+	UserName   string
+	BaseURL    string
+	HTTPClient *http.Client
+}
+
+// NewWikipediaClient knows how to create a new client.
+// The client impements NameResolver interface.
+func NewWikipediaClient(username string) (*WikipediaClient, error) {
+	if username == "" {
+		return nil, errors.New("nil username")
+	}
+	c := WikipediaClient{
+		UA:       userAgent,
+		UserName: username,
+		BaseURL:  baseURLGeoNames,
+		HTTPClient: &http.Client{
+			Timeout: 5 * time.Second,
+		},
+	}
+	return &c, nil
+}
+
+// GetCoordinates knows how to retrive geo coordinates for
+// the given place name and country code.
+func (w WikipediaClient) GetCoordinates(placeName, country string) (Place, error) {
+	u, err := w.makeURL(placeName)
+	if err != nil {
+		return Place{}, err
+	}
+	req, err := prepareRequest(u)
+	if err != nil {
+		return Place{}, err
+	}
+	res, err := w.HTTPClient.Do(req)
+	if err != nil {
+		return Place{}, err
+	}
+	defer res.Body.Close()
+
+	var wp wikipediaPlaces
+	if err := json.NewDecoder(res.Body).Decode(&wp); err != nil {
+		return Place{}, fmt.Errorf("decoding response %w", err)
```

Comment:
I suggest reading all the body with `io.ReadAll` and then passing the resulting `[]byte` to `json.Unmarshal`. In the event of a parsing error, it'll be nice to have the JSON data to display to the user. Using the `Decode` API, we wouldn't have that, because consuming the reader destroys its contents.

Reply:
implemented